### PR TITLE
add support for deeply resolving properties

### LINKED ISSUE #55: withProp should recursively resolve prop getters.
I ended up finding a solution to my problem while writing this issue, but I believe it's something that styled-tools should handle by default. I'm working with the sample below which a reakit theme, and the issue is in `contrastText`. What ends up happening is that `bg` resolves to a prop-getter function instead of a string. The reason being that `bgNeedle` is a prop-getter that resolves to a prop-getter which is not supported by `withProp`. What I ended up doing is writing a custom implementation of `withProp` (below the example) that recursively resolves prop-getter functions. It would be useful if this was either the default behavior of `withProp` or was exposed as a separate function.

```ts
// utils.ts

import { range } from 'lodash-es';
import { getLuminance, lighten } from 'polished';
import { Needle, palette as p, withProp } from 'styled-tools';

export const contrastText = (bgNeedle: Needle<any>) =>
  withProp(
    [bgNeedle, p('black'), p('white')] as any,
    (bg: string, black: string, white: string) => {
      // bg ends up being a prop getter function here instead of a string.
      return getLuminance(bg) > 0.179 ? black : white;
    },
  );

export const contrastPalette = (palette: string, tone?: number) =>
  contrastText(p(palette, tone));

export const lightenPalette = (
  amount: number,
  palette: string,
  tone?: number,
) =>
  withProp([p(palette, tone)] as any, (color: string) =>
    lighten(amount, color),
  );

export const tonePalette = (palette: string) => [
  p(palette),
  ...range(0.1, 0.5, 0.1).map(i => lightenPalette(i, palette)),
];

export const tonePaletteText = (palette: string) =>
  range(5).map(i => contrastPalette(palette, i));

export const tonePaletteWithText = (name: string, palette: string) => ({
  [name]: tonePalette(palette),
  [`${name}Text`]: tonePaletteText(name)
});
```

```ts
// index.ts

import { tonePaletteWithText } from './utils';

export const palette = {
  white: '#fff',
  black: '#000',

  astronautBlue: '#003a5d',

  ...tonePaletteWithText('primary', 'astronautBlue'),
};
```

## Solution

```ts
// simple deepWithProp

const deepWithProp = <Props extends object, T = undefined>(
  needle: Needle<Props> | Needle<Props>[],
  fn: (...args: any[]) => T
) => (props = {}): T => {
  if (Array.isArray(needle)) {
    const needles = needle.map(arg => deepWithProp(arg, x => x)(props));
    return fn(...needles);
  }
  if (typeof needle === "function") {
    return fn(deepWithProp(needle(props as Props), x => x)(props));
  }
  return fn(prop(needle, needle)(props));
};
```

### PR DESCRIPTION
**Pull request by jtmthf**
Fixes #55. All functions now allow for result values to be deeply resolved allowing for more complex expressions. One such use case for these deeply resolved properties is themes with multiple layers of dependent properties.

### CODE REVIEW

**Comment by diegohaz** [comment]
Thank you so much @jtmthf. Could you please rebase your branch with current master (upgraded dependencies so the following command will work) and copy and paste here the benchmark results?
```
yarn benchmark ifNotProp
yarn benchmark ifProp
yarn benchmark prop
yarn benchmark switchProp
yarn benchmark withProp
```

**Comment by jtmthf** [comment]
I added another fix as `palette` was not resolving functions that returned arrays. As part of that, I added a section of the code I'm working on jtmthf/reakit-theme-bootstrap as a test case. Let me know if you don't need the added complexity of that test and would prefer something more generic.

This all now allows the following code to work:

```ts
// utils.ts

import { palette as p, theme as t, Needle, withProp } from "styled-tools";
import { getLuminance, tint } from "polished";
import { range } from "lodash-es";

export const contrastText = (bgNeedle: Needle<any>) =>
  withProp(bgNeedle, (bg: string) =>
    getLuminance(bg) > 0.179 ? p("black") : p("white")
  );

export const contrastPalette = (key: string, tone?: number) =>
  contrastText(p(key, tone));

export const tintPalette = (amount: number, key: string, tone?: number) =>
  withProp(p(key, tone) as any, (color: string) => tint(amount, color));

export const tonePalette = (key: string, tone?: number) =>
  withProp(t("colorInterval") as any, (colorInterval: number) =>
    range(5)
      .map(i => i * colorInterval)
      .map(i => tintPalette(i, key, tone))
  );

export const tonePaletteText = (key: string) =>
  range(5).map(i => contrastPalette(key, i));

export const tonePaletteWithText = (
  name: string,
  key: string,
  tone?: number
) => ({
  [name]: tonePalette(key, tone),
  [`${name}Text`]: tonePaletteText(name)
});
```

```ts
// index.ts

import { palette as p } from "styled-tools";
import { tonePaletteWithText } from "./utils";

export const palette = {
  white: "#fff",
  gray100: "#f8f9fa",
  gray200: "#e9ecef",
  gray300: "#dee2e6",
  gray400: "#ced4da",
  gray500: "#adb5bd",
  gray600: "#6c757d",
  gray700: "#495057",
  gray800: "#343a40",
  gray900: "#212529",
  black: "#000",

  grays: [
    p("gray100"),
    p("gray200"),
    p("gray300"),
    p("gray400"),
    p("gray500"),
    p("gray600"),
    p("gray700"),
    p("gray800"),
    p("gray900")
  ],

  blue: "#007bff",
  indigo: "#6610f2",
  purple: "#6f42c1",
  pink: "#e83e8c",
  red: "#dc3545",
  orange: "#fd7e14",
  yellow: "#ffc107",
  green: "#28a745",
  teal: "#20c997",
  cyan: "#17a2b8",

  ...tonePaletteWithText("primary", "blue"),
  ...tonePaletteWithText("secondary", "grays", 5),
  ...tonePaletteWithText("success", "green"),
  ...tonePaletteWithText("info", "cyan"),
  ...tonePaletteWithText("warning", "yellow"),
  ...tonePaletteWithText("danger", "red"),
  ...tonePaletteWithText("light", "grays"),
  ...tonePaletteWithText("dark", "grays", -2)
};

export const colorInterval = 0.08;
```

```mdx
---
name: Box
menu: components
route: /components/box
---

import { Playground, PropsTable } from "docz";
import { Box, Flex } from "reakit";
import { range } from "lodash-es";

<Playground>
  <Flex>
    {[
      "primary",
      "secondary",
      "success",
      "info",
      "warning",
      "danger",
      "light",
      "dark"
    ].map(palette => (
      <Flex key={palette} column flex={1}>
        {range(5).map(tone => (
          <Box key={tone} palette={palette} opaque tone={tone}>
            {palette}
          </Box>
        ))}
      </Flex>
    ))}
  </Flex>
</Playground>
```

which renders to:

Reakit Bootstrap

**Comment by jtmthf** [comment]
@diegohaz is anything still needed to merge this?

**Comment by diegohaz** [comment]
Sorry for not answering before @jtmthf. I still didn't find the time to test it properly. Since this is a big change, I wouldn't like to ship it during the holidays because I wouldn't have time to fix bugs that can possibly appear. 

In the meanwhile, it'll help if you could use your fork on your project so as to catch possible regressions.

**Comment by jtmthf** [comment]
@diegohaz no problem, I'm not releasing anything for a couple of weeks. I'll keep testing in the meantime with the fork.

File: src/theme.js [hunk 1]
```diff
@@ -14,6 +14,6 @@ import type { PropsWithTheme } from ".";
  * `;
  */
 const theme = (path: string, defaultValue?: any) => (props: PropsWithTheme) =>
-  prop(path, defaultValue)(props.theme);
+  prop(`theme.${path}`, defaultValue)(props);
```
Comment (diegohaz) [inline src/theme.js]: Why this change?
